From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Christian Gmeiner <cgmeiner@igalia.com>
Date: Mon, 25 Aug 2025 12:33:06 +0200
Subject: [PATCH] util/perf: Fix sysprof TLS destructor crash on library unload

When Mesa is compiled with sysprof support, applications can crash with a
segfault during shutdown. This happens because sysprof_collector_mark()
registers thread-local storage destructors that get called after the library
containing the destructor code has been unloaded.

Fix this by adding the NODELETE linker flag when sysprof is enabled,
preventing drivers that use mesa_util from being unloaded and keeping
the TLS destructors valid.

Fixes: 9da5eafa8ec ("util/perf: Add sysprof integration")
CC: mesa-stable
Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/13571
Signed-off-by: Christian Gmeiner <cgmeiner@igalia.com>
---
 src/util/meson.build | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/util/meson.build b/src/util/meson.build
index bf67bf16692a..235d63e0bcaf 100644
--- a/src/util/meson.build
+++ b/src/util/meson.build
@@ -278,37 +278,40 @@ if with_gpuvis
   )
 endif
 
+mesa_util_link_args = []
 if with_sysprof
   files_mesa_util += files(
     'perf/u_sysprof.c',
     'perf/u_sysprof.h',
   )
   deps_for_libmesa_util += dep_sysprof
+  mesa_util_link_args += ['-Wl,-z,nodelete']
 endif
 
 u_trace_py = files('perf/u_trace.py')
 
 libmesa_util_simd = static_library(
   'mesa_util_simd',
   files('streaming-load-memcpy.c'),
   c_args : [c_msvc_compat_args, sse41_args],
   include_directories : [inc_util],
   gnu_symbol_visibility : 'hidden',
   build_by_default : false,
 )
 
 _libmesa_util = static_library(
   'mesa_util',
   [files_mesa_util, files_debug_stack, format_srgb],
   include_directories : [inc_util, include_directories('format')],
   dependencies : deps_for_libmesa_util,
   link_with: [libmesa_util_simd],
   c_args : [c_msvc_compat_args],
   gnu_symbol_visibility : 'hidden',
   build_by_default : false
 )
 
 idep_mesautil = declare_dependency(
+  link_args : mesa_util_link_args,
   link_with : _libmesa_util,
   include_directories : [inc_util],
   dependencies : deps_for_libmesa_util,
